pkg/app/pipedv1/livestatereporter/livestatereporter.go

```diff
@@ -45,127 +51,226 @@ type Reporter interface {
+	if err != nil {
+		rlogger.Error("failed to create working directory", zap.Error(err))
+		return nil, err
+	}
+
 	r := &reporter{
-		reporters: make([]pluginReporter, 0, len(plugins)),
-		logger:    rlogger,
-	}
-
-	for name, p := range plugins {
-		r.reporters = append(r.reporters, pluginReporter{
-			pluginName:            name,
-			snapshotFlushInterval: time.Minute,
-			appLister:             appLister,
-			apiClient:             apiClient,
-			pluginClient:          p,
-			logger:                rlogger.With(zap.String("plugin-name", name)),
-		})
+		snapshotFlushInterval: 1 * time.Minute,
+		appLister:             appLister,
+		apiClient:             apiClient,
+		gitClient:             gitClient,
+		repoMap:               make(map[string]git.Repo),
+		pluginRegistry:        pluginRegistry,
+		pipedConfig:           pipedConfig,
+		secretDecrypter:       secretDecrypter,
+		workingDir:            workingDir,
+		logger:                rlogger,
 	}
-	return r
+
+	return r, nil
 }
 
 func (r *reporter) Run(ctx context.Context) error {
-	group, ctx := errgroup.WithContext(ctx)
+	r.logger.Info("start running app live state reporter", zap.Duration("snapshot-flush-interval", r.snapshotFlushInterval))
 
-	r.logger.Info(fmt.Sprintf("starting %d live state reporters of plugins", len(r.reporters)))
+	snapshotTicker := time.NewTicker(r.snapshotFlushInterval)
+	defer snapshotTicker.Stop()
 
-	for _, reporter := range r.reporters {
-		// Avoid starting all reporters at the same time to reduce the API call burst.
-		time.Sleep(10 * time.Second)
+	for {
+		select {
+		case <-snapshotTicker.C:
+			r.flushSnapshots(ctx)
 
-		group.Go(func() error {
-			reporter.logger.Info("starting app live state reporter for plugin")
-			return reporter.Run(ctx)
-		})
+		case <-ctx.Done():
+			r.logger.Info("app live state reporter has been stopped")
+			return nil
+		}
 	}
+}
 
-	r.logger.Info(fmt.Sprintf("all live state reporters of %d plugins have been started", len(r.reporters)))
+func (r *reporter) flushSnapshots(ctx context.Context) {
+	r.logger.Info("start flushing snapshots")
 
-	if err := group.Wait(); err != nil {
-		r.logger.Error("failed while running", zap.Error(err))
-		return err
+	r.logger.Info("updating git repositories")
+	for id, cfgRepo := range r.pipedConfig.GetRepositoryMap() {
+		if _, ok := r.repoMap[id]; ok {
+			r.logger.Info("the repo is cloned, so pulling repository", zap.String("repo-id", id))
+			err := r.repoMap[id].Pull(ctx, cfgRepo.Branch)
+			if err != nil {
+				r.logger.Error("failed to pull repository", zap.String("repo-id", id), zap.Error(err))
+			}
+			continue
+		}
+
+		r.logger.Info("the repo is not cloned, so cloning repository", zap.String("repo-id", id))
+		repo, err := r.gitClient.Clone(ctx, id, cfgRepo.Remote, cfgRepo.Branch, fmt.Sprintf("%s/%s", r.workingDir, id))
+		if err != nil {
+			r.logger.Error("failed to clone repository", zap.String("repo-id", id), zap.Error(err))
+			continue
+		}
+		defer repo.Clean()
```

Comment:
The `repo.Clean()` deletes the cloned directory.
So if we reuse the git.Repo, we must not call `repo.Clean()`
https://github.com/pipe-cd/pipecd/blob/8ce525249e2e6c9f74c3df8417e8cc0aabcdfe06/pkg/git/repo.go#L360-L363

By the way, we make a local cache when `gitClient.Clone`, so it's not so expensive to clone every time. If you want to update the source cloner every time, I think we can make gitSourceCloner every time instead of using localSourceCloner.

Reply:
@Warashi 

Thanks, removed `repo.Clean()` on f7bef7f67.

> By the way, we make a local cache when gitClient.Clone, so it's not so expensive to clone every time. If you want to update the source cloner every time, I think we can make gitSourceCloner every time instead of using localSourceCloner.

I checked the logic and noticed that the `Clone` method executes `git fetch` if the repo cache is found.
https://github.com/pipe-cd/pipecd/blob/786bf3de6794a0f6b80cd524b946bf149d5540f8/pkg/git/client.go#L180C1-L195C4

I want to execute `git pull` to update the branch's HEAD and use the commit every time.
So I want to use localSourceCloner instead of gitSourceCloner 🙏

Reply:
Okay, I got it.